File: health_advanced/HealthDashboard/Controllers/HDTabBarController.m
//
//  HDTabBarController.m
//  HealthDashboard
//
//  Created by zhang, haizi on 2026/3/19.
//

#import "HDTabBarController.h"
#import "HDDashboardViewController.h"
#import "HDExerciseTypeViewController.h"
#import "HDProfileViewController.h"
#import "../Models/HDHealthDataModel.h"

@implementation HDTabBarController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setupTabs];
    [self applyTabBarStyle];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(themeChanged)
                                                 name:@"HDThemeDidChange"
                                               object:nil];
}

- (void)setupTabs {
    // 首页
    HDDashboardViewController *dash = [HDDashboardViewController new];
    dash.extendedLayoutIncludesOpaqueBars = YES;
    dash.edgesForExtendedLayout = UIRectEdgeAll;
    UINavigationController *dashNav = [[UINavigationController alloc] initWithRootViewController:dash];
    dashNav.navigationBarHidden = YES;
    dashNav.extendedLayoutIncludesOpaqueBars = YES;
    dashNav.edgesForExtendedLayout = UIRectEdgeAll;
    dashNav.tabBarItem = [[UITabBarItem alloc] initWithTitle:@"首页"
                                                       image:[UIImage systemImageNamed:@"heart.fill"]
                                                         tag:0];
    
    // 运动
    HDExerciseTypeViewController *exercise = [HDExerciseTypeViewController new];
    UINavigationController *exerciseNav = [[UINavigationController alloc] initWithRootViewController:exercise];
    exerciseNav.navigationBarHidden = NO;
    exerciseNav.tabBarItem = [[UITabBarItem alloc] initWithTitle:@"运动"
                                                           image:[UIImage systemImageNamed:@"figure.walk"]
                                                             tag:1];
    
    // 个人中心
    HDProfileViewController *profile = [HDProfileViewController new];
    UINavigationController *profNav  = [[UINavigationController alloc] initWithRootViewController:profile];
    profNav.navigationBarHidden = YES;
    profNav.tabBarItem = [[UITabBarItem alloc] initWithTitle:@"我的"
                                                       image:[UIImage systemImageNamed:@"person.circle.fill"]
                                                         tag:2];
    self.viewControllers = @[dashNav, exerciseNav, profNav];
}

- (void)applyTabBarStyle {
    BOOL dark = [HDHealthDataModel shared].isDarkMode;
    UITabBar *tb = self.tabBar;
    if (@available(iOS 15, *)) {
        UITabBarAppearance *app = [UITabBarAppearance new];
        [app configureWithOpaqueBackground];
        app.backgroundColor = UIColor.systemBackgroundColor;
        tb.standardAppearance   = app;
        tb.scrollEdgeAppearance = app;
    } else {
        tb.barTintColor = UIColor.systemBackgroundColor;
        tb.translucent = NO;
    }
    tb.tintColor               = UIColor.systemBlueColor;
    tb.unselectedItemTintColor = UIColor.secondaryLabelColor;
}

- (void)themeChanged {
    [self applyTabBarStyle];
    for (UINavigationController *nav in self.viewControllers) {
        if ([nav.topViewController respondsToSelector:@selector(applyTheme)]) {
            [(id)nav.topViewController applyTheme];
        }
    }
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

@end

File: health_advanced_swift/HealthDashboard/Controllers/HDTabBarController.swift
//
//  HDTabBarController.swift
//  HealthDashboard - TabBar 控制器
//
//  Created by zhang, haizi on 2026/3/22.
//

import UIKit
import Combine

@MainActor
class HDTabBarController: UITabBarController {

    private var cancellables = Set<AnyCancellable>()

    // MARK: - Lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()
        setupTabs()
        applyTabBarStyle()
        // 订阅主题变化
        HDHealthDataModel.shared.$isDarkMode
            .receive(on: DispatchQueue.main)
            .sink { [weak self] _ in self?.applyTabBarStyle() }
            .store(in: &cancellables)
    }

    // MARK: - Tab Setup

    private func setupTabs() {
        // Tab 0: 首页
        let dashboard = HDDashboardViewController()
        dashboard.extendedLayoutIncludesOpaqueBars = true
        dashboard.edgesForExtendedLayout = .all
        let dashNav = UINavigationController(rootViewController: dashboard)
        dashNav.isNavigationBarHidden = true
        dashNav.extendedLayoutIncludesOpaqueBars = true
        dashNav.edgesForExtendedLayout = .all
        dashNav.tabBarItem = UITabBarItem(title: "首页", image: UIImage(systemName: "heart.fill"), tag: 0)

        // Tab 1: 运动
        let exercise = HDExerciseTypeViewController()
        let exerciseNav = UINavigationController(rootViewController: exercise)
        exerciseNav.isNavigationBarHidden = false
        exerciseNav.tabBarItem = UITabBarItem(title: "运动", image: UIImage(systemName: "figure.walk"), tag: 1)

        // Tab 2: 个人中心
        let profile = HDProfileViewController()
        let profileNav = UINavigationController(rootViewController: profile)
        profileNav.isNavigationBarHidden = true
        profileNav.tabBarItem = UITabBarItem(title: "我的", image: UIImage(systemName: "person.circle.fill"), tag: 2)

        viewControllers = [dashNav, exerciseNav, profileNav]
    }

    // MARK: - Theme

    private func applyTabBarStyle() {
        let tb = tabBar
        if #available(iOS 15, *) {
            let app = UITabBarAppearance()
            app.configureWithOpaqueBackground()
            app.backgroundColor = .systemBackground
            tb.standardAppearance = app
            tb.scrollEdgeAppearance = app
        } else {
            tb.barTintColor = .systemBackground
            tb.isTranslucent = false
        }
        tb.tintColor = .systemBlue
        tb.unselectedItemTintColor = .secondaryLabel
    }

    @objc private func themeChanged() {
        applyTabBarStyle()
        // 各 VC 通过自己监听 HDThemeDidChange 通知响应主题切换
    }
}
